File: openapi-test-java-server-jaxrs/testapi/api/Responses.java
package testapi.api;

import com.fasterxml.jackson.annotation.JsonProperty;
import java.lang.Void;
import testapi.model.Error;

/** Response type for: 201, 400 */
public sealed interface Response201400<T201, T400> permits Created, BadRequest {
  @JsonProperty("status")
  String status();
}

/** Response type for: 404, default */
public sealed interface Response404Default<T404> permits NotFound, Default {
  @JsonProperty("status")
  String status();
}

/** Response type for: 200, 404 */
public sealed interface Response200404<T200, T404> permits Ok, NotFound {
  @JsonProperty("status")
  String status();
}

/** Response type for: 200, 4XX, 5XX */
public sealed interface Response2004XX5XX<T200> permits Ok, ClientError4XX, ServerError5XX {
  @JsonProperty("status")
  String status();
}

/** HTTP 400 response */
public record BadRequest<T>(@JsonProperty("value") T value) implements Response201400<Void, T> {
  public BadRequest<T> withValue(T value) {
    return new BadRequest<>(value);
  };

  @Override
  public String status() {
    return "400";
  };
}

/** HTTP 201 response */
public record Created<T>(@JsonProperty("value") T value) implements Response201400<T, Void> {
  public Created<T> withValue(T value) {
    return new Created<>(value);
  };

  @Override
  public String status() {
    return "201";
  };
}

/** HTTP default response */
public record Default(
  /** HTTP status code */
  @JsonProperty("statusCode") Integer statusCode,
  @JsonProperty("value") Error value
) implements Response404Default<Void> {
  /** HTTP status code */
  public Default withStatusCode(Integer statusCode) {
    return new Default(statusCode, value);
  };

  public Default withValue(Error value) {
    return new Default(statusCode, value);
  };

  @Override
  public String status() {
    return "default";
  };
}

/** HTTP 5XX response */
public record ServerError5XX(
  /** HTTP status code */
  @JsonProperty("statusCode") Integer statusCode,
  @JsonProperty("value") Error value
) implements Response2004XX5XX<Void> {
  /** HTTP status code */
  public ServerError5XX withStatusCode(Integer statusCode) {
    return new ServerError5XX(statusCode, value);
  };

  public ServerError5XX withValue(Error value) {
    return new ServerError5XX(statusCode, value);
  };

  @Override
  public String status() {
    return "5XX";
  };
}

/** HTTP 200 response */
public record Ok<T>(@JsonProperty("value") T value) implements Response200404<T, Void>, Response2004XX5XX<T> {
  public Ok<T> withValue(T value) {
    return new Ok<>(value);
  };

  @Override
  public String status() {
    return "200";
  };
}

/** HTTP 404 response */
public record NotFound<T>(@JsonProperty("value") T value) implements Response404Default<T>, Response200404<Void, T> {
  public NotFound<T> withValue(T value) {
    return new NotFound<>(value);
  };

  @Override
  public String status() {
    return "404";
  };
}

/** HTTP 4XX response */
public record ClientError4XX(
  /** HTTP status code */
  @JsonProperty("statusCode") Integer statusCode,
  @JsonProperty("value") Error value
) implements Response2004XX5XX<Void> {
  /** HTTP status code */
  public ClientError4XX withStatusCode(Integer statusCode) {
    return new ClientError4XX(statusCode, value);
  };

  public ClientError4XX withValue(Error value) {
    return new ClientError4XX(statusCode, value);
  };

  @Override
  public String status() {
    return "4XX";
  };
}

File: openapi-test-kotlin-server-spring/testapi/api/Responses.kt
package testapi.api

import com.fasterxml.jackson.annotation.JsonProperty
import kotlin.Nothing
import testapi.model.Error

/** Response type for: 201, 400 */
sealed interface Response201400<out T201, out T400> {
  @JsonProperty("status")
  fun status(): String
}

/** Response type for: 200, 404 */
sealed interface Response200404<out T200, out T404> {
  @JsonProperty("status")
  fun status(): String
}

/** Response type for: 200, 4XX, 5XX */
sealed interface Response2004XX5XX<out T200> {
  @JsonProperty("status")
  fun status(): String
}

/** HTTP 400 response */
data class BadRequest<out T>(@field:JsonProperty("value") val value: T) : Response201400<Nothing, T> {
  override fun status(): String = "400"
}

/** HTTP 201 response */
data class Created<out T>(@field:JsonProperty("value") val value: T) : Response201400<T, Nothing> {
  override fun status(): String = "201"
}

/** HTTP 5XX response */
data class ServerError5XX(
  /** HTTP status code */
  @field:JsonProperty("statusCode") val statusCode: Int,
  @field:JsonProperty("value") val value: Error
) : Response2004XX5XX<Nothing> {
  override fun status(): String = "5XX"
}

/** HTTP 200 response */
data class Ok<out T>(@field:JsonProperty("value") val value: T) : Response200404<T, Nothing>, Response2004XX5XX<T> {
  override fun status(): String = "200"
}

/** HTTP 404 response */
data class NotFound<out T>(@field:JsonProperty("value") val value: T) : Response200404<Nothing, T> {
  override fun status(): String = "404"
}

/** HTTP 4XX response */
data class ClientError4XX(
  /** HTTP status code */
  @field:JsonProperty("statusCode") val statusCode: Int,
  @field:JsonProperty("value") val value: Error
) : Response2004XX5XX<Nothing> {
  override fun status(): String = "4XX"
}